daffodil-runtime2/src/main/resources/org/apache/daffodil/runtime2/c/libruntime/parsers.c

```diff
@@ -18,140 +18,484 @@
-            char           c_val[sizeof(type##bits##_t)];                                                    \
-            type##bits##_t i_val;                                                                            \
-        } buffer;                                                                                            \
-                                                                                                             \
-        read_stream_update_position(&buffer.c_val, sizeof(type##bits##_t));                                  \
-        *number = endian##bits##toh(buffer.i_val);                                                           \
-    }
-
-// Parse binary booleans, real numbers, and integers
-
-define_parse_endian_bool(be, 16)
-define_parse_endian_bool(be, 32)
-define_parse_endian_bool(be, 8)
-
-define_parse_endian_real(be, double, 64)
-define_parse_endian_real(be, float, 32)
-
-define_parse_endian_integer(be, int, 16)
-define_parse_endian_integer(be, int, 32)
-define_parse_endian_integer(be, int, 64)
-define_parse_endian_integer(be, int, 8)
-
-define_parse_endian_integer(be, uint, 16)
-define_parse_endian_integer(be, uint, 32)
-define_parse_endian_integer(be, uint, 64)
-define_parse_endian_integer(be, uint, 8)
-
-define_parse_endian_bool(le, 16)
-define_parse_endian_bool(le, 32)
-define_parse_endian_bool(le, 8)
-
-define_parse_endian_real(le, double, 64)
-define_parse_endian_real(le, float, 32)
-
-define_parse_endian_integer(le, int, 16)
-define_parse_endian_integer(le, int, 32)
-define_parse_endian_integer(le, int, 64)
-define_parse_endian_integer(le, int, 8)
-
-define_parse_endian_integer(le, uint, 16)
-define_parse_endian_integer(le, uint, 32)
-define_parse_endian_integer(le, uint, 64)
-define_parse_endian_integer(le, uint, 8)
-
-// Parse fill bytes until end position is reached
-
-void
-parse_fill_bytes(size_t end_position, PState *pstate)
+// Helper macros to get "n" highest bits from a byte's high end
+
+#define BYTE_WIDTH 8
+#define BIG_ENDIAN_DATA 1
+#define LITTLE_ENDIAN_DATA 0
+#define LOW_MASK(n) ((1 << n) - 1)
+#define HIGH_MASK(n) (LOW_MASK(n) << (BYTE_WIDTH - n))
+#define HIGH_BITS(byte, n) ((byte & HIGH_MASK(n)) >> (BYTE_WIDTH - n))
+
+// Helper method to read bits using whole bytes while storing
+// remaining bits not yet read within a fractional byte; returns last
+// bits of last byte in normal position (NOT shifted left)
+
+static void
+read_bits(uint8_t *bytes, size_t num_bits, PState *pstate)
+{
+    // Copy as many bytes directly from stream as possible
+    size_t ix_bytes = 0;
+    if (!pstate->unreadLen)
+    {
+        size_t num_bytes = num_bits / BYTE_WIDTH;
+        if (num_bytes)
+        {
+            size_t count = fread(bytes, 1, num_bytes, pstate->stream);
+            if (count < num_bytes)
+            {
+                pstate->error = eof_or_error(pstate->stream);
+                return;
+            }
+            pstate->bitPos0b += count * BYTE_WIDTH;
```

Comment:
As I think about this some more, I wonder if `read_bits` doesn't want to update bitPos0b until after it has successfully read the full num_bits, i.e. right before the function returns success?

For example, if we're at bit position X and we need Y bits for some number, if we can't read all Y bits then we probably want a diagnostic that says "Y bits were not available starting at bit position X". But if we increase bitPos0 before we finish reading all bits and then hit an error, the error message would look like "Y bits were not available starting at position Z", where Z is some position after X, and it might not make much sense.

So I'd argue bitPosition wants to be the position the last place where read_bits succeeded (even if we don't actually output those diagnostics like that). Which simplifies the logic, the lastline of the function just becomes `pstate->bitPos0b += num_bits` and it's the only place bitPos0b is changed in the function.

There might even be an argument that the bitPosition shouldn't be updated until the read bits are successfully converted to a number. That way if there was an error you could say something like "failed to convert X bits to a boolean, starting at position Y". Again, I don't know if runtime2 outputs that kind of diagnostic, but it makes me think that that is what bitPos wants to represent: the last position of a successful read.